azurerm/resource_arm_monitor_diagnostics.go

```diff
@@ -0,0 +1,284 @@
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+func resourceArmMonitorDiagnostics() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceArmMonitorDiagnosticsCreateUpdate,
+		Read:   resourceArmMonitorDiagnosticsRead,
+		Update: resourceArmMonitorDiagnosticsCreateUpdate,
+		Delete: resourceArmMonitorDiagnosticsDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"target_resource_id": {
+				Type:             schema.TypeString,
+				Required:         true,
+				ForceNew:         true,
+				DiffSuppressFunc: ignoreCaseDiffSuppressFunc,
+			},
+
+			"storage_account_id": {
+				Type:             schema.TypeString,
+				Optional:         true,
+				DiffSuppressFunc: ignoreCaseDiffSuppressFunc,
+			},
+
+			"event_hub_name": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"event_hub_authorization_rule_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"workspace_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"disabled_settings": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+
+			"retention_days": {
+				Type:     schema.TypeInt,
+				Optional: true,
+			},
+		},
+	}
+}
+
+func resourceArmMonitorDiagnosticsCreateUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).monitorDiagnosticSettingsClient
+	ctx := meta.(*ArmClient).StopContext
+	log.Printf("[INFO] preparing arguments for Azure ARM Diagnostic Settings.")
+
+	name := d.Get("name").(string)
+	targetResourceId := d.Get("target_resource_id").(string)
+	storageAccountId := d.Get("storage_account_id").(string)
+	eventHubName := d.Get("event_hub_name").(string)
+	eventHubAuthorizationRuleId := d.Get("event_hub_authorization_rule_id").(string)
+	workspaceId := d.Get("workspace_id").(string)
+	disabledSettings := d.Get("disabled_settings").([]interface{})
+	retentionDays := d.Get("retention_days").(int)
+
+	allMetricSettings, allLogSettings, err := getAllDiagnosticSettings(targetResourceId, meta)
+	if err != nil {
+		return err
+	}
+
+	allSettings := append(*allMetricSettings, *allLogSettings...)
+	if !utils.StringSliceContainsStringSlice(disabledSettings, allSettings) {
+		return fmt.Errorf("Invalid value for disabled settings provided, use one or multiple of: %q", allSettings)
+	}
+
+	if len(allSettings) == len(disabledSettings) {
+		return fmt.Errorf("You can not disable all settings, rather delete diagnostic logging")
+	}
+
+	diagnosticSettings := &insights.DiagnosticSettings{}
+	diagnosticSettings.Metrics = expandMetricsConfiguration(*allMetricSettings, disabledSettings, retentionDays)
+	diagnosticSettings.Logs = expandLogConfiguration(*allLogSettings, disabledSettings, retentionDays)
```

Comment:
so I need to look into this further, but I think `metrics` and `logs` want to be nested elements here, rather than having one setting to configure all of the values. Whilst I appreciate this makes it simpler in some circumstances (e.g. when you want to enable everything) - this means the values could change between Terraform runs which would be unexpected, should a new log become available. Admittedly the same thing is true for a Data Source, but they're intended for Dynamic Data (so this wouldn't be out of place)